### PR TITLE
[BMQ] Add log with session/queue operations time

### PR DESCRIPTION
**Pull request by alexander-e1off**
Augment the logs of operation such as start/stop session as well as open/configure/close queue (both sync async versions) to contain the human friendly (for easy log read) as well as the nanoseconds time of all operations (in both success and failures).
Log looks like this:
```
INFO bmqimp_brokersession.cpp:1391 Open queue [uri=bmq://bmq.test.persistent.priority/my-first-queue] took: 18.38 ms (18383531 nanoseconds)

INFO bmqimp_brokersession.cpp:940 Stop took: 3.66 ms (3662015 nanoseconds)
```

### CODE REVIEW

File: src/groups/bmq/bmqimp/bmqimp_brokersession.h [hunk 1]
```diff
@@ -377,6 +377,9 @@ class BrokerSession BSLS_CPP11_FINAL {
         bsl::vector<StateTransition> d_transitionTable;
         // State transition table
 
+        bsls::Types::Int64 d_beginTimestamp;
+        // HiRes timer value of the begin start/stop operation
```
Comment (678098) [inline src/groups/bmq/bmqimp/bmqimp_brokersession.h:381]:
```suggestion
        /// HiRes timer value of the begin start/stop operation
        bsls::Types::Int64 d_beginTimestamp;
```

File: src/groups/bmq/bmqimp/bmqimp_brokersession.h [hunk 2]
```diff
@@ -408,6 +411,9 @@ class BrokerSession BSLS_CPP11_FINAL {
         /// if not void indicates state entry logic execution result.
         void setClosingChannel(FsmEvent::Enum event);
 
+        void logOperationTime(const char* operation);
+        /// Log start/stop operation time.
```
Comment (678098) [inline src/groups/bmq/bmqimp/bmqimp_brokersession.h:415]:
```suggestion
        /// Log start/stop operation time for the specified `operation`,
        /// using the stored operation begin timestamp.
        /// Reset the begin timestamp to 0.
        void logOperationTime(const char* operation);
```

File: src/groups/bmq/bmqimp/bmqimp_brokersession.h [hunk 3]
```diff
@@ -461,6 +467,10 @@ class BrokerSession BSLS_CPP11_FINAL {
         bsl::vector<QueueStateTransition> d_transitionTable;
         // State transition table
 
+        typedef bsl::unordered_map<int, bsls::Types::Int64> TimestampMap;
```
Comment (678098) [inline src/groups/bmq/bmqimp/bmqimp_brokersession.h:470]: Should go to TYPES section

File: src/groups/bmq/bmqimp/bmqimp_brokersession.h [hunk 4]
```diff
@@ -581,6 +591,9 @@ class BrokerSession BSLS_CPP11_FINAL {
         /// Initiate the resumption of a queue.
         void actionInitiateQueueResume(const bsl::shared_ptr<Queue>& queue);
 
+        /// Log queue operation time
```
Comment (678098) [inline src/groups/bmq/bmqimp/bmqimp_brokersession.h:594]:
```suggestion
        /// Log start/stop operation time for the specified `queue` and `operation`,
        /// using the stored operation begin timestamp.
        /// Reset the begin timestamp to 0.
```

File: src/groups/bmq/bmqimp/bmqimp_brokersession.cpp [hunk 5]
```diff
@@ -919,6 +932,17 @@ void BrokerSession::SessionFsm::handleAllQueuesResumed()
     }
 }
 
+void BrokerSession::SessionFsm::logOperationTime(const char* operation)
+{
+    if (d_beginTimestamp) {
```
[the remark this answers is not in the document]
Comment (alexander-e1off) [reply src/groups/bmq/bmqimp/bmqimp_brokersession.cpp:937]: Added negative branch with warning.

File: src/groups/bmq/bmqimp/bmqimp_brokersession.cpp [hunk 6]
```diff
@@ -919,6 +932,17 @@ void BrokerSession::SessionFsm::handleAllQueuesResumed()
...
+        const bsls::Types::Int64 elapsed =
+            mwcsys::Time::highResolutionTimer() - d_beginTimestamp;
+        BALL_LOG_INFO << operation << " took: "
+                      << mwcu::PrintUtil::prettyTimeInterval(elapsed) << " ("
```
[the remark this answers is not in the document]
Comment (alexander-e1off) [reply src/groups/bmq/bmqimp/bmqimp_brokersession.cpp:941]: Agree, it is used in non-critical performance paths.

File: src/groups/bmq/bmqimp/bmqimp_brokersession.cpp [hunk 7]
```diff
@@ -1350,8 +1381,25 @@ void BrokerSession::QueueFsm::actionInitiateQueueResume(
         context);
 }
 
+void BrokerSession::QueueFsm::logOperationTime(const int   queueId,
```
Comment (678098) [inline src/groups/bmq/bmqimp/bmqimp_brokersession.cpp:1384]: I am curious if logging `queueId` here is a good way.
We might log queue uri instead, and it's easier to search in logs.
Queue uri for the same queue will always be the same across different launches and even across client/cluster, but `queueId` might be different.

File: src/groups/bmq/bmqimp/bmqimp_brokersession.h [hunk 8]
```diff
@@ -581,6 +598,11 @@ class BrokerSession BSLS_CPP11_FINAL {
         /// Initiate the resumption of a queue.
         void actionInitiateQueueResume(const bsl::shared_ptr<Queue>& queue);
 
+        /// Log start/stop/configure operation time for the specified `queue`
```
Comment (678098) [inline src/groups/bmq/bmqimp/bmqimp_brokersession.h:601]:
```suggestion
        /// Log start/stop/configure operation time for the specified `queueId`
```

**Review by 678098 — COMMENTED** [review]
Last questions

File: src/groups/bmq/bmqimp/bmqimp_brokersession.cpp (hunk 7 above)
Comment (alexander-e1off) [reply src/groups/bmq/bmqimp/bmqimp_brokersession.cpp]: Inside `BrokerSession` class `queueId` is used for logging states transitions, but agree with you that for operation time `queueUri` is more readable/searchable. Changed to `queueUri`.

**Review by 678098 — APPROVED** [review]
LGTM

**Pull request merged**